bookserver/routers/assessment.py

```diff
@@ -76,3 +79,63 @@ async def get_assessment_results(
     #       res.update(res_update)
     rslogger.debug(f"Returning {row}")
     return make_json_response(detail=row)
+
+
+# Define a simple model for the gethist request.
+# If you just try to specify the two fields as parameters it expects
+# them to be in a query string.
+class HistoryRequest(BaseModel):
+    acid: str
+    sid: Optional[str] = None
+
+
+@router.post("/gethist")
+async def get_history(request: Request, request_data: HistoryRequest):
+    """
+    return the history of saved code by this user for a particular acid
+    :Parameters:
+        - `acid`: id of the active code block
+        - `user`: optional identifier for the owner of the code
+    :Return:
+        - json object with a detail key that references a dictionary
+
+        ::
+
+            { "acid": div_id,
+              "sid" : id of student requested,
+              "history": [code, code, code],
+              "timestamps": [ts, ts, ts]
+            }
+    """
+    acid = request_data.acid
+    sid = request_data.sid
+    # if request_data.sid then we know this is being called from the grading interface
+    # so verify that the actual user is an instructor.
+    if sid:
+        if request.state.user and request.state.user.username != sid:
+            if await is_instructor(request):
+                course_id = request.state.user.course_id
+            else:
+                course_id = None
+        else:
+            raise HTTPException(401)
+
+    elif request.state.user:
+        sid = request.state.user.username
+        course_id = request.state.user.course_id
+    else:
+        sid = None
+        course_id = None
+
+    res: Dict[str, Any] = {}
+    if sid:
```

Comment:
Will sid or acid ever be null if the else clause isn't taken? Meaning, is this test necessary, or should it be an assert?